crates/provider/src/layers/gas.rs

```diff
@@ -0,0 +1,367 @@
+///
+/// # Example
+///
+/// ```rs
+/// # async fn test<T: Transport + Clone, S: NetworkSigner<Ethereum>>(transport: T, signer: S) {
+/// let provider = ProviderBuilder::new()
+///     .layer(ManagedNonceLayer)
+///     .layer(GasEstimatorLayer)
+///     .signer(EthereumSigner::from(signer)) // note the order!
+///     .provider(RootProvider::new(transport));
+///
+/// provider.send_transaction(TransactionRequest::default()).await;
+/// # }
+#[derive(Debug, Clone, Copy)]
+pub struct GasEstimatorLayer;
+
+impl<P, N, T> ProviderLayer<P, N, T> for GasEstimatorLayer
+where
+    P: Provider<N, T>,
+    N: Network,
+    T: Transport + Clone,
+{
+    type Provider = GasEstimatorProvider<N, T, P>;
+    fn layer(&self, inner: P) -> Self::Provider {
+        GasEstimatorProvider { inner, _phantom: PhantomData }
+    }
+}
+
+/// A provider that estimates gas for transactions.
+///
+/// Note: This provider requires the chain_id to be set in the transaction request if it's a
+/// EIP1559.
+///
+/// You cannot construct this directly, use [`ProviderBuilder`] with a [`GasEstimatorLayer`].
+///
+/// [`ProviderBuilder`]: crate::ProviderBuilder
+#[derive(Debug, Clone)]
+pub struct GasEstimatorProvider<N, T, P>
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<N, T>,
+{
+    inner: P,
+    _phantom: PhantomData<(N, T)>,
+}
+
+impl<N, T, P> GasEstimatorProvider<N, T, P>
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<N, T>,
+{
+    /// Gets the gas_price to be used in legacy txs.
+    async fn get_gas_price(&self) -> TransportResult<U256> {
+        self.inner.get_gas_price().await
+    }
+
+    /// Gets the gas_limit to be used in txs.
+    async fn get_gas_estimate(&self, tx: &N::TransactionRequest) -> TransportResult<U256> {
+        self.inner.estimate_gas(tx, None).await
+    }
+
+    /// Gets the max_fee_per_gas and max_priority_fee_per_gas to be used in EIP-1559 txs.
+    async fn get_eip1559_fees_estimate(&self) -> TransportResult<(U256, U256)> {
+        self.inner.estimate_eip1559_fees(None).await
+    }
+
+    /// Populates the gas_limit, max_fee_per_gas and max_priority_fee_per_gas fields if unset.
+    /// Requires the chain_id to be set in the transaction request to be processed as a EIP-1559 tx.
+    /// If the network does not support EIP-1559, it will process it as a legacy tx.
+    async fn handle_eip1559_tx(
+        &self,
+        mut tx: N::TransactionRequest,
+    ) -> Result<N::TransactionRequest, TransportError> {
+        // Check if the user has set the gas_limit, max_fee_per_gas and max_priority_fee_per_gas
+        // fields.
+        let gas_limit_unset = tx.gas_limit().is_none();
+        // eip1559_fees_unset is true if any one of the two max_fee_per_gas or
+        // max_priority_fee_per_gas is unset.
+        let eip1559_fees_unset =
+            tx.max_fee_per_gas().is_none() || tx.max_priority_fee_per_gas().is_none();
+
+        match (gas_limit_unset, eip1559_fees_unset) {
```

Comment:
remove the combinatoric match please

follow the pattern in `populate_gas` and `populate_gas_eip1559` in `provider.rs`